File: app/src/main/java/mono/hg/wrappers/TextSpectator.java
package mono.hg.wrappers;

import android.os.Handler;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

import java.lang.ref.WeakReference;

/**
 * TextSpectator is an extension to TextWatcher, used when it is attached to an EditText.
 */
public class TextSpectator implements TextWatcher {
    // The default duration of the Handler loop, in milliseconds.
    private static int DEFAULT_TICK_DURATION = 125;

    private WeakReference<EditText> watchField;
    private int tickDuration = DEFAULT_TICK_DURATION;
    private boolean timerStopped = false;
    private boolean spaceSpam = false;

    private Handler handler = new Handler();
    private Runnable runnable = new Runnable() {
        @Override public void run() {
            if (!timerStopped) {
                whenTimerTicked();
                handler.postDelayed(this, tickDuration);
            }
        }
    };

    protected TextSpectator(EditText editText) {
        watchField = new WeakReference<>(editText);
    }

    /**
     * Fetch the input text of the attached EditText;
     *
     * @return The contents of EditText.
     */
    protected String getInputText() {
        if (watchField != null) {
            return watchField.get().getText().toString();
        } else {
            return null;
        }
    }

    /**
     * Fetch the input text of the attached EditText. Trimmed using String.trim().
     *
     * @return The trimmed contents of EditText;
     */
    public String getTrimmedInputText() {
        if (watchField != null) {
            return watchField.get().getText().toString().trim();
        } else {
            return null;
        }
    }

    /**
     * Should the user be allowed to start the query with empty spaces?
     *
     * @param enabled Whether query can start with an empty space.
     */
    public void spaceSpamming(boolean enabled) {
        spaceSpam = enabled;
    }

    /**
     * Stops the Handler loop.
     */
    protected void stopTimer() {
        timerStopped = true;
    }

    /**
     * Starts the Handler loop.
     */
    protected void startTimer() {
        timerStopped = false;
        handler.postDelayed(runnable, tickDuration);
    }

    /**
     * Sets the Handler loop duration.
     *
     * @param duration In milliseconds, how long should the Handler delay before looping?
     */
    public void setTimerDuration(int duration) {
        tickDuration = duration;
    }

    /**
     * Actions done when the Handler completes a single loop.
     */
    public void whenTimerTicked() {
        // Overridden when necessary.
    }

    public void beforeChanged(CharSequence s, int start, int count, int after) {
        // Overridden when necessary.
    }

    public void whenChanged(CharSequence s, int start, int before, int count) {
        // Overridden when necessary.
    }

    public void afterChanged(Editable s) {
        // Overridden when necessary.
    }

    @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        beforeChanged(s, start, count, after);
    }

    @Override public void onTextChanged(CharSequence s, int start, int before, int count) {
        whenChanged(s, start, before, count);
    }

    @Override public void afterTextChanged(Editable s) {
        if (!spaceSpam && s.length() > 0 && s.charAt(0) == ' ') {
            s.delete(0, 1);
        }

        afterChanged(s);
    }
}

File: app/src/main/java/mono/hg/wrappers/TextSpectator.kt
package mono.hg.wrappers

import android.os.Handler
import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import java.lang.ref.WeakReference

/**
 * TextSpectator is an extension to TextWatcher, used when it is attached to an EditText.
 */
open class TextSpectator protected constructor(editText: EditText) : TextWatcher {
    private val watchField: WeakReference<EditText>?
    private var tickDuration = DEFAULT_TICK_DURATION
    private var timerStopped = false
    private val handler = Handler()
    private val runnable: Runnable = object : Runnable {
        override fun run() {
            if (! timerStopped) {
                whenTimerTicked()
                handler.postDelayed(this, tickDuration.toLong())
            }
        }
    }

    /**
     * Fetch the input text of the attached EditText;
     *
     * @return The contents of EditText.
     */
    protected val inputText: String?
        get() = if (watchField != null) {
            watchField.get() !!.text.toString()
        } else {
            null
        }

    /**
     * Fetch the input text of the attached EditText. Trimmed using String.trim().
     *
     * @return The trimmed contents of EditText;
     */
    val trimmedInputText: String
        get() = if (watchField != null) {
            watchField.get() !!.text.toString().trim { it <= ' ' }
        } else {
            ""
        }

    /**
     * Stops the Handler loop.
     */
    protected fun stopTimer() {
        timerStopped = true
    }

    /**
     * Starts the Handler loop.
     */
    protected fun startTimer() {
        timerStopped = false
        handler.postDelayed(runnable, tickDuration.toLong())
    }

    /**
     * Sets the Handler loop duration.
     *
     * @param duration In milliseconds, how long should the Handler delay before looping?
     */
    fun setTimerDuration(duration: Int) {
        tickDuration = duration
    }

    /**
     * Actions done when the Handler completes a single loop.
     */
    open fun whenTimerTicked() {
        // Overridden when necessary.
    }

    /**
     * A syntactic sugar for beforeTextChanged().
     */
    fun beforeChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        // Overridden when necessary.
    }

    /**
     * A syntactic sugar for onTextChanged().
     */
    open fun whenChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        // Overridden when necessary.
    }

    /**
     * A syntactic sugar for afterTextChanged().
     */
    open fun afterChanged(s: Editable?) {
        // Overridden when necessary.
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        beforeChanged(s, start, count, after)
    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        whenChanged(s, start, before, count)
    }

    override fun afterTextChanged(s: Editable) {
        if (s.isNotEmpty() && s[0] == ' ') {
            s.delete(0, 1)
        }
        afterChanged(s)
    }

    companion object {
        // The default duration of the Handler loop, in milliseconds.
        private const val DEFAULT_TICK_DURATION = 125
    }

    init {
        watchField = WeakReference(editText)
    }
}